Sources/NIO/ConvenienceOptionSupport.swift

```diff
@@ -0,0 +1,440 @@
+        if let updatedUnderlying = underlyingBootstrap.applyChannelOption(option) {
+            return NIOClientTCPBootstrap(self, withUpdated: updatedUnderlying)
+        } else {
+            let applier = NIOClientTCPBootstrap_Applier(contained: self)
+            return option.applyOptionDefaultMapping(with: applier).contained
+        }
+    }
+    
+    struct NIOClientTCPBootstrap_Applier : NIOChannelOptionAppliable {
+        var contained : NIOClientTCPBootstrap
+        
+        func applyOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> NIOClientTCPBootstrap_Applier {
+            return NIOClientTCPBootstrap_Applier(contained: contained.channelOption(option, value: value))
+        }
+    }
+}
+
+// MARK: Utility
+/// An object which can have a 'ChannelOption' applied to it and will return an appropriately updated version of itself.
+public protocol NIOChannelOptionAppliable {
+    /// Apply a ChannelOption and return an updated version of self.
+    func applyOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> Self
+}
+
+// MARK: TCP - data
+/// A TCP channel option which can be applied to a bootstrap or similar using shorthand notation.
+public struct NIOTCPShorthandOption  {
+    private var data: ShorthandOption
+    
+    private init(_ data: ShorthandOption) {
+        self.data = data
+    }
+    
+    /// Apply the contained option to the supplied object (almost certainly bootstrap) using the default mapping.
+    /// - Parameter with: object to use to apply the option.
+    /// - Returns: the modified object
+    public func applyOptionDefaultMapping<OptionApplier : NIOChannelOptionAppliable>(with optionApplier: OptionApplier) -> OptionApplier {
+        return data.applyOptionDefaultMapping(with: optionApplier)
+    }
+    
+    fileprivate enum ShorthandOption {
+        case reuseAddr
+        case disableAutoRead
+        case allowRemoteHalfClosure(Bool)
+        
+        func applyOptionDefaultMapping<OptionApplier : NIOChannelOptionAppliable>(with optionApplier: OptionApplier) -> OptionApplier {
+            switch self {
+            case .reuseAddr:
+                return optionApplier.applyOption(ChannelOptions.socketOption(.reuseaddr), value: 1)
+            case .allowRemoteHalfClosure(let value):
+                return optionApplier.applyOption(ChannelOptions.allowRemoteHalfClosure, value: value)
+            case .disableAutoRead:
+                return optionApplier.applyOption(ChannelOptions.autoRead, value: false)
+            }
+        }
+    }
+}
+
+// Hashable for the convenience of users.
+extension NIOTCPShorthandOption: Hashable {}
+extension NIOTCPShorthandOption.ShorthandOption: Hashable {}
+
+/// Approved shorthand options.
+extension NIOTCPShorthandOption {
+    /// Allow immediately reusing a local address.
+    public static let allowImmediateEndpointAddressReuse = NIOTCPShorthandOption(.reuseAddr)
```

Comment:
I think we should add the word `Local` to the `allowImmediate...`. (I'm aware I proposed this name but I think I missed the "Local" bit).

@glbrntt / @Lukasa any good ideas?

Reply:
Added local.